scripts/deleteSnapsets.js

```diff
@@ -0,0 +1,162 @@
+            aSnapset.value._deleted = true;
+            dbLoader.docsToRemove.push(aSnapset.value);
+        });
+        dbLoader.addGpiiKeysAndBulkDelete(dbLoader.snapSets.rows, dbLoader.docsToRemove);
+    });
+};
+
+/**
+ * Find the GPII key records associated with the given snapset Prefs Safes,
+ * delete each one as found, and then batch delete all the snapset Prefs Safes.
+ * @param {Array} snapSets      - The snapsets of interest.
+ * @param {Array} docsToRemove  - Array of records to delete.
+ */
+dbLoader.addGpiiKeysAndBulkDelete = function (snapSets, docsToRemove) {
+    fluid.each(snapSets, function (aSnapset) {
+        var gpiiKeyId = aSnapset.value._id;
+        fluid.log("Snapset: " + gpiiKeyId);
+        var gpiiKeyViewUrl = dbLoader.gpiiKeyViewUrl + "?key=%22" + gpiiKeyId + "%22";
+        var getGpiiKeysRequest = http.request(gpiiKeyViewUrl, function (resp) {
+            var respString = "";
+            resp.setEncoding("utf8");
+            resp.on("data", function (chunk) {
+                respString += chunk;
+            });
+            // This response "end" event doesn't finish until after the call to
+            // doBatchDelete() below, which is too late to add the GPII Key to
+            // docsToRemove and have them removed.
+            // Note - if there is a way to determine the final "end" event, then
+            // it could  call doBatchDelete().
+            resp.on("end", function () {
+                var gpiiKeyRecords = JSON.parse(respString);
+                fluid.each(gpiiKeyRecords.rows, function (gpiiKey) {
+                    fluid.log("GPII Key: " + gpiiKey.value._id);
+                    dbLoader.deleteGpiiKey(gpiiKey.value);
+                });
+            });
+        });
+        getGpiiKeysRequest.on("error", function (e) {
+            fluid.log("Error finding snapsets' associated GPII Keys: " + e.message);
+        });
+        getGpiiKeysRequest.end();
+    });
+    dbLoader.doBatchDelete(docsToRemove);
+};
+
+/**
+ * Delete the given GPII Key record.
+ * @param {Object} gpiiKey  - The key to delete.
+ */
+dbLoader.deleteGpiiKey = function (gpiiKey) {
+    var deleteOptions = {
+        hostname: dbLoader.parsedCouchDbUrl.hostname,
+        port: dbLoader.parsedCouchDbUrl.port,
+        path: "",           // filled in below.
+        method: "DELETE",
+        headers: {
+            "Accept": "application/json"
+        }
+    };
+    deleteOptions.path = "/gpii/" + gpiiKey._id + "?rev=" + gpiiKey._rev;
+    var deleteGpiiKeyRequest = http.request(deleteOptions, function (res) {
+        fluid.log("STATUS: " + res.statusCode);
+        fluid.log("HEADERS: " + JSON.stringify(res.headers, null, 2));
+    });
+    deleteGpiiKeyRequest.on("error", function (e) {
+        fluid.log("Error finding GPII Key: " + e.message);
+    });
+    deleteGpiiKeyRequest.end();
+};
+
+/**
+ * Delete the snapset Prefs Safes and their associated GPII Keys.
```

Comment:
This function only deletes prefs safes, not gpii keys.

Reply:
With the use of promises, this now does delete all prefs safes and associated gpii keys.